package/utils/mdadm/files/mdadm.init

```diff
@@ -71,14 +99,14 @@ mdadm_array() {
 }
 
 start_service() {
-	local email
-
 	mkdir -p "${CONF%/*}"
 	printf "# Autogenerated from /etc/config/mdadm, do not edit!\n" > $CONF
 
-	config_load mdadm
-	config_foreach mdadm_common mdadm
-	config_foreach mdadm_array array
+	if ! mdadm_autoconfig; then
+		config_load mdadm
+		config_foreach mdadm_common mdadm
+		config_foreach mdadm_array array
+	fi
 
 	$PROG --assemble --scan --config="$CONF"
```

Comment:
Can you make CONF optional?

Reply:
will comment more later (but kudos for all the work you're doing on this).

Thist is what I'm referring to though '--assemble with --scan' is special-cased to do:

```
   Auto Assembly
       When  --assemble is used with --scan and no devices are listed, mdadm will first attempt to assemble
       all the arrays listed in the config file.

       If no arrays are listed in the config (other than those marked <ignore>) it will  look  through  the
       available devices for possible arrays and will try to assemble anything that it finds.  Arrays which
       are tagged as belonging to the given homehost will be assembled and started normally.  Arrays  which
       do not obviously belong to this host are given names that are expected not to conflict with anything
       local, and are started "read-auto" so that nothing is written to any device until the array is writ‐
       ten to. i.e.  automatic resync etc is delayed.
```

Reply:
Although I'm think that would be best with an empty config file, so skipping --conf bit isn't really necessary (I think back when I did it, I relied on shipping an entirely commented (or non-existent) mdadm.conf file so that the kernel would auto assemble.  However that was based on a fairly linear setup and doesn't deal with e.g. hotplug which is the norm in openwrt (in my case I was using old x86 hardware with internal hard drives).

Reply:
I think _--config=none_ would produce the desired results with minimal effort.  It would be fairly trivial to tweak `$CONF=none` rather than rely on _/var/etc/mdadm.conf_.  It's certainly worth testing.

Introducing a second uci defaults option increases complexity and flexibility, e.g. `option config none` could pass through the actual (valid) values of **none**, **partition**, or (writable) **file**.  If `config` is unset or empty then stay using the global value of _/var/etc/mdadm.conf_.  If `autoconfig` is true (or false) **and** `config` is 'none' then attempt stateless Auto Assembly regardless? Or, should it always prefer to use a stateful config if `autoconfig` is true and `config` is 'none'?

Another option would be to use just one `config` option with 3 (or more) validated values and some would pass through.  Valid values would be **auto**, **partition**, **none**, or a (writable) **file** location.

Reply:
I'm not a core dev (was at one point, but 'resigned' as I was not well.  Thing are much better, but I haven't tried to get back in...also not really doing much with the core system these days), so take what I say with a gain of salt, but my preference is the latter.  I think the use is much more clear. @jow @blogic thoughts?